types/react/index.d.ts

```diff
@@ -359,55 +359,156 @@ declare namespace React {
     // Component Specs and Lifecycle
     // ----------------------------------------------------------------------
 
-    interface ComponentLifecycle<P, S> {
+    // This should actually be something like `Lifecycle<P, S> | DeprecatedLifecycle<P, S>`,
+    // as React will _not_ call the deprecated lifecycle methods if any of the new lifecycle
+    // methods are present.
+    interface ComponentLifecycle<P, S> extends NewLifecycle<P, S>, DeprecatedLifecycle<P, S> {
+        /**
+         * Called immediately after a compoment is mounted. Setting state here will trigger re-rendering.
+         */
+        componentDidMount?(): void;
+        /**
+         * Called to determine whether the change in props and state should trigger a re-render.
+         *
+         * `Component` always returns true.
+         * `PureComponent` implements a shallow comparison on props and state and returns true if any
+         * props or states have changed.
+         *
+         * If false is returned, `Component#render`, `componentWillUpdate`
+         * and `componentDidUpdate` will not be called.
+         */
+        shouldComponentUpdate?(nextProps: Readonly<P>, nextState: Readonly<S>, nextContext: any): boolean;
+        /**
+         * Called immediately before a component is destroyed. Perform any necessary cleanup in this method, such as
+         * cancelled network requests, or cleaning up any DOM elements created in `componentDidMount`.
+         */
+        componentWillUnmount?(): void;
+        /**
+         * Catches exceptions generated in descendant components. Unhandled exceptions will cause
+         * the entire component tree to unmount.
+         */
+        componentDidCatch?(error: Error, errorInfo: ErrorInfo): void;
+    }
+
+    // Unfortunately, we have no way of declaring that the component constructor must implement this
+    interface StaticLifecycle<P, S> {
+        getDerivedStateFromProps?: GetDerivedStateFromProps<P, S>;
+    }
+
+    type GetDerivedStateFromProps<P, S> =
+        /**
+         * Returns an update to a component's state based on its new props and old state.
+         *
+         * Note: its presence prevents any of the deprecated lifecycle methods from being invoked
+         */
+        <K extends keyof S>(nextProps: Readonly<P>, prevState: Readonly<S>) => Pick<S, K> | S | null;
```

Comment:
since this in reality an abstract static method, having the generic type parameter `K` doesn't actually help the developer with inferring what they need to return (they instead need to provide the parser the keys they want to pick from state), we could just leave this as `(nextProps: Readonly<P>, prevState: Readonly<S>) => Partial<S> | null`

otherwise doing this will cause a type error
```ts
interface State {
  a: string
  b: number
}

const foo: StaticLifecycle<{}, State> = {
  getDerivedStateFromProps(nextProps, prevState) { // nextProps and prevState inferred without type annotations
    return {
      b: 0 // error because K is not inferred, nor should it really need to be
    }
  }
}
```

Reply:
This was an attempt to preserve the `setState` behavior of allowing one to omit keys but not allowing one to return `undefined` as a value if it's not actually in the domain of the value.

It can't really be expressed in TypeScript without requiring generics hacks 😭

Reply:
Ah yeah, good point, that's unfortunate...